raphtory/src/algorithms/motifs/temporal_rich_club_coefficient.rs

```diff
@@ -0,0 +1,181 @@
+use std::{
+    borrow::Borrow,
+    cmp::{self, max, min},
+    collections::HashSet,
+    default,
+};
+
+use raphtory_api::core::entities::VID;
+use rustc_hash::FxHashSet;
+
+use crate::prelude::{EdgeViewOps, GraphViewOps, NodeViewOps};
+
+struct SlidingWindows<I> {
+    iter: I,
+    window_size: usize,
+}
+
+impl<I> SlidingWindows<I> {
+    fn new(iter: I, window_size: usize) -> Self {
+        SlidingWindows { iter, window_size }
+    }
+}
+
+impl<I> Iterator for SlidingWindows<I>
+where
+    I: Iterator,
+    I::Item: Clone,
+{
+    type Item = Vec<I::Item>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut window = Vec::with_capacity(self.window_size);
+        for _ in 0..self.window_size {
+            if let Some(item) = self.iter.next() {
+                window.push(item);
+            } else {
+                return None;
+            }
+        }
+        Some(window)
+    }
+}
+
+pub fn temporal_rich_club_coefficient<'a, I, G1, G2>(
+    agg_graph: G2,
+    views: I,
+    k: usize,
+    window_size: usize,
+) -> f64
+where
+    I: IntoIterator<Item = G1>,
+    G1: GraphViewOps<'a>,
+    G2: GraphViewOps<'a>,
+{
+    // Extract the set of nodes with degree greater than or equal to k
+    let s_k: HashSet<VID> = agg_graph
+        .nodes()
+        .into_iter()
+        .filter(|v| v.degree() >= k)
+        .map(|v| v.node)
+        .collect();
+
+    let temp_rich_club_val = SlidingWindows::new(views.into_iter(), window_size)
+        .map(|window| intermediate_rich_club_coef(s_k.clone(), window))
+        .reduce(f64::max)
+        .unwrap_or(0.0);
+
+    temp_rich_club_val
+}
+
+fn intermediate_rich_club_coef<'a, I, G1>(s_k: HashSet<VID>, views: I) -> f64
+where
+    I: IntoIterator<Item = G1>,
+    G1: GraphViewOps<'a>,
+{
+    // Extract the edges among the top degree nodes which are stable over that subset of snapshots by computing their intersection
+    let stable_edges = views
+        .into_iter()
+        .map(|g| {
+            let new_edges: HashSet<UndirEdge> = g
+                .subgraph(s_k.clone())
+                .edges()
+                .into_iter()
+                .map(|e| undir_edge(e.src().node, e.dst().node))
+                .collect();
+            new_edges
+        })
+        .into_iter()
+        .reduce(|acc_edges, item_edges| acc_edges.intersection(&item_edges).cloned().collect());
```

Comment:
possibly worth it to use `into_par_iter` here?

Reply:
Thank you! I'll have a look now.